app/Services/VersionService.php

```diff
@@ -90,15 +75,68 @@ public function publishNewVersionFromGithub(string $tag)
             throw new VersionAlreadyExistsException();
         }
 
+        try {
+            $normalisedVersion = $this->versionParser->normalize($tag);
+        } catch (UnexpectedValueException) {
+            Log::error(sprintf('Invalid release channel %s', $tag));
+            throw new ReleaseChannelNotFoundException();
+        }
+
+        $releaseChannel = PluginReleaseChannel::where('name', VersionParser::parseStability($normalisedVersion))
+            ->first();
+
+        if (!$releaseChannel) {
+            Log::error(sprintf('Invalid release channel %s', $tag));
+            throw new ReleaseChannelNotFoundException();
+        }
+
         // Create the version
         $newVersion = Version::create(
             [
-                'version' => $tag
+                'version' => $tag,
+                'plugin_release_channel_id' => $releaseChannel->id,
             ]
         );
 
         // Retire old versions
-        Version::where('id', '<>', $newVersion->id)->delete();
+        Version::where('id', '<>', $newVersion->id)
+            ->get()
+            ->reject(function (Version $version) use ($newVersion) {
+                return Comparator::greaterThan(
+                    $version->version,
+                    $newVersion->version
+                );
+            })
+            ->each(function (Version $version) {
+                $version->delete();
+            });
+    }
+
+    /**
+     * The "latest version" on any release channel is the most recent version (in semver terms)
+     * on any channel that is, or more stable than, the requested channel.
+     */
+    public function getLatestVersionForReleaseChannel(string $channel): Version
+    {
+        $relevantVersions = $this->getRelevantVersions(PluginReleaseChannel::where('name', $channel)->firstOrFail());
+        if ($relevantVersions->isEmpty()) {
+            throw new VersionNotFoundException();
+        }
+
+        return $relevantVersions->reduce(function (?Version $selectedVersion, Version $version) {
+            return is_null($selectedVersion) || Comparator::greaterThan(
+                $version->version,
+                $selectedVersion->version
+            ) ? $version : $selectedVersion;
+        });
+    }
+
+    private function getRelevantVersions(PluginReleaseChannel $requestedChannel): Collection
+    {
+        return Version::whereHas('pluginReleaseChannel', function (Builder $releaseChannel) use ($requestedChannel) {
+            return $releaseChannel->where('relative_stability', '<=', $requestedChannel->relative_stability);
+        })
+            ->get();
```

Comment:
Seems a little bit odd on its own line :D

Reply:
Made it less lonely